### PR TITLE
Use a single thread for all ConnectionPool Reapers

### PR DESCRIPTION
**Pull request by jhawthorn**
Previously we would spawn one thread per connection pool, which ends up being wasteful for apps with many connection pools, since the reaping frequency is usually the same.

The savings from this aren't actually that significant, in my testing each thread only uses ~16k memory (`ObjectSpace.memsize_of` reports 1M, but I think that's inaccurate), but I think it's still worth doing (it makes debugging easier if nothing else).

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb [hunk 1]
```diff
@@ -294,15 +294,33 @@ def initialize(pool, frequency)
...
+        def self.register_pool(pool, frequency) # :nodoc:
+          @@mutex.synchronize do
+            if @@pools.key?(frequency)
+              @@pools[frequency] << pool
+            else
+              @@pools[frequency] = [pool]
+              Thread.new(frequency) do |t|
+                loop do
+                  sleep t
+                  @@mutex.synchronize do
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb]: Why the mutex here? It seems to be that we any shared resource?

Comment (jhawthorn) [reply activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb]: We're [iterating over `@@pools[frequency]`](https://github.com/rails/rails/pull/36296/files#diff-7735c67f539e7d1e2908a4d5d8909c24R310), which is also modified [here](https://github.com/rails/rails/pull/36296/files#diff-7735c67f539e7d1e2908a4d5d8909c24R303)

**Comment by eileencodes** [comment]
Backported to 6.0

**Comment by aeroastro** [comment]
Thank you very much for fixing this issue ❤️

In our environment, when we upgraded to Rails 5.2 from Rails 4, we once suffered from the resource exhaustion where thread count (not memory size) of the Reaper over multiple databases and multiple processes actually reached kernel's limitation. The issue then was resolved by a workaround to disable Reaper, but this PR resolves the issue fundamentally.

We're looking forward to upgrading to Rails 6, and thank you again for your great work 🎉

**Pull request merged**